server/schedulers/hot_region_v2.go

```diff
@@ -0,0 +1,369 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The v2 selection algorithm related code is placed in this file.
+
+package schedulers
+
+import (
+	"fmt"
+	"math"
+)
+
+const (
+	firstPriorityPerceivedRatio = 0.2  // PeerRate needs to be 20% above what needs to be balanced.
+	firstPriorityMinHotRatio    = 0.02 // PeerRate needs to be greater than 1% lowRate
+
+	secondPriorityPerceivedRatio = 0.3  // PeerRate needs to be 20% above what needs to be balanced.
+	secondPriorityMinHotRatio    = 0.03 // PeerRate needs to be greater than 1.5% lowRate
+)
+
+// isAvailable returns the solution is available.
+// If the solution has no revertRegion, progressiveRank should < 0.
+// If the solution has some revertRegion, progressiveRank should equal to -4 or -3.
+func isAvailableV2(s *solution) bool {
+	// TODO: Test if revert region can be enabled for -1.
+	return s.progressiveRank <= -3 || (s.progressiveRank < 0 && s.revertRegion == nil)
+}
+
+// TODO: Unified with stddevThreshold.
+type rankV2Ratios struct {
+	preBalancedRatio      float64
+	balancedRatio         float64
+	preBalancedCheckRatio float64
+	balancedCheckRatio    float64
+	perceivedRatio        float64
+	minHotRatio           float64
+}
+
+func newRankV2Ratios(balancedRatio, perceivedRatio, minHotRatio float64) *rankV2Ratios {
+	// limit 0.7 <= balancedRatio <= 0.95
+	if balancedRatio < 0.7 {
+		balancedRatio = 0.7
+	}
+	if balancedRatio > 0.95 {
+		balancedRatio = 0.95
+	}
+
+	rs := &rankV2Ratios{balancedRatio: balancedRatio, perceivedRatio: perceivedRatio, minHotRatio: minHotRatio}
+	rs.preBalancedRatio = math.Max(2.0*balancedRatio-1.0, balancedRatio-0.15)
```

Comment:
why not use `balancedRatio-0.15` directly :), it's hard to understand the purpose... (0.85 is the intersection)

Reply:
The original formula is `1.0 - 2 * (1.0 - balancedRatio)`. The maximum value with `balancedRatio-0.1` is to prevent the preBalance range becoming too large.